### PR TITLE
Merging dynamic shapes feature branch.

### PR DESCRIPTION
**Pull request by benvanik**
This implements some new VM/tblgen support to enable more advanced marshaling (such as lists of tuples, strings, etc). These features are used to emit nice layout type ops and a much nicer centralized layout/executable caching scheme. Finally, the layouts are extended to support push constants and marshal primitive types to executables.

This also switches i32->index almost everywhere, as it was the only way to make all of the new code dealing with shape dimensions and offsets and descriptor set info work.

### CODE REVIEW

File: iree/modules/hal/hal_module.cc [hunk 1]
```diff
@@ -555,19 +558,18 @@ class HALModuleState final {
...
     absl::InlinedVector<iree_hal_descriptor_set_binding_t, 4> binding_structs(
-        bindings.size());
-    for (int i = 0; i < bindings.size(); ++i) {
+        binding_ordinals.size());
+    for (int i = 0; i < binding_ordinals.size(); ++i) {
       binding_structs[i] = {
-          std::get<0>(bindings[i]), std::get<1>(bindings[i]).get(),
-          static_cast<iree_device_size_t>(std::get<2>(bindings[i])),
-          static_cast<iree_device_size_t>(std::get<3>(bindings[i]))};
+          binding_ordinals[i], binding_buffers[i].get(),
```
Comment (stellaraccident) [inline iree/modules/hal/hal_module.cc]: eol comments for each would help legibility (and also keep clang-format from globbing together)

File: iree/compiler/Dialect/VM/IR/VMOps.td [hunk 2]
```diff
@@ -795,6 +795,79 @@ def VM_SelectRefOp : VM_PureOp<"select.ref", [
...
 }
 
+class VM_SwitchPrimitiveOp<Type type, string mnemonic, VM_OPC opcode,
+                           list<OpTrait> traits = []> :
+    VM_PureOp<mnemonic, !listconcat(traits, [
+      DeclareOpInterfaceMethods<VM_SerializableOpInterface>,
+      AllTypesMatch<["default_value", "result"]>,
+    ])> {
+  let description = [{
+    Returns the value with the given `index` in `values` or `default_value` if
```
Comment (stellaraccident) [inline iree/compiler/Dialect/VM/IR/VMOps.td]: Please provide an ASM example

File: iree/compiler/Dialect/VM/IR/VMOps.td [hunk 3]
```diff
@@ -795,6 +795,79 @@ def VM_SelectRefOp : VM_PureOp<"select.ref", [
...
+  let arguments = (ins
+    I32:$index,
+    I32:$default_value,
+    Variadic<type>:$values
+  );
+  let results = (outs
+    type:$result
+  );
+
+  // let assemblyFormat = [{
```
Comment (stellaraccident) [inline iree/compiler/Dialect/VM/IR/VMOps.td]: TODO?

File: iree/compiler/Dialect/VM/IR/VMOps.td [hunk 4]
```diff
@@ -795,6 +795,79 @@ def VM_SelectRefOp : VM_PureOp<"select.ref", [
...
+  let hasFolder = 1;
+}
+
+def VM_SwitchRefOp : VM_PureOp<"switch.ref", [
+    DeclareOpInterfaceMethods<VM_SerializableOpInterface>,
+    AllTypesMatch<["default_value", "result"]>,
+  ]> {
+  let summary = [{ref_ptr switch operation}];
+  let description = [{
+    Returns the value with the given `index` in `values` or `default_value` if
```
Comment (stellaraccident) [inline iree/compiler/Dialect/VM/IR/VMOps.td]: "See example in ..."

File: iree/compiler/Dialect/VM/IR/VMOps.td [hunk 5]
```diff
@@ -795,6 +795,79 @@ def VM_SelectRefOp : VM_PureOp<"select.ref", [
...
+  let arguments = (ins
+    I32:$index,
+    VM_AnyRef:$default_value,
+    Variadic<VM_AnyRef>:$values
+  );
+  let results = (outs
+    VM_AnyRef:$result
+  );
+
+  // let assemblyFormat = [{
```
Comment (stellaraccident) [inline iree/compiler/Dialect/VM/IR/VMOps.td]: TODO

File: iree/compiler/Dialect/VM/IR/test/assignment_folding.mlir [hunk 6]
```diff
@@ -58,3 +58,42 @@ vm.module @select_ref_folds {
...
+    %c0 = vm.const.i32 0 : i32
+    %c1 = vm.const.i32 1 : i32
+    %c2 = vm.const.i32 2 : i32
+    %c3 = vm.const.i32 3 : i32
+    %c100 = vm.const.i32 100 : i32
+    %c200 = vm.const.i32 200 : i32
+    %c300 = vm.const.i32 300 : i32
+    %c400 = vm.const.i32 400 : i32
+    // CHECK: vm.return %c100, %c200, %c300, %c400 : i32, i32, i32, i32
+    %0 = vm.switch.i32 [%c100, %c200, %c300](%c0) else %c400 : i32
```
Comment (stellaraccident) [inline iree/compiler/Dialect/VM/IR/test/assignment_folding.mlir]: Nit: I find this syntax hard to parse. Reversing it makes more sense to me:
(%c0)[...] else %...

File: iree/vm/bytecode_dispatch.c [hunk 7]
```diff
@@ -494,6 +496,59 @@ iree_status_t iree_vm_bytecode_dispatch(
...
+      //   VM_EncVariadicOperands<"values">,
+      //   VM_EncResult<"result">,
+      // ];
+      int32_t index = OP_R_I32(0);
+      int32_t type_id = OP_I32(2);
+      iree_vm_ref_t* default_value = &OP_R_REF(6);
+      int is_move = OP_R_REF_IS_MOVE(6);
+      const iree_vm_register_list_t* src_reg_list =
+          (const iree_vm_register_list_t*)&bytecode_data[offset + 8];
+      offset +=
```
Comment (stellaraccident) [inline iree/vm/bytecode_dispatch.c]: I feel like this line could use a comment, such as "advance to skip ..."

Comment (stellaraccident) [inline iree/vm/bytecode_dispatch.c]: Meta-point not related to this pr: I wish that "offset" were called something to signify that it is a global/part of VM state (i.e. we are advancing the instruction pointer). I mis-read "offset" as a local variable often.

File: iree/compiler/Dialect/Flow/Transforms/test/transformation.mlir [hunk 8]
```diff
@@ -189,13 +189,15 @@ func @dynamicUpdateSlice(%operand : tensor<2x4xi32>, %update : tensor<1x1xi32>,
 // CHECK-NEXT:   }
 // CHECK-NEXT: }
 // CHECK-NEXT: func @dynamicUpdateSlice(%arg0: tensor<2x4xi32>, %arg1: tensor<1x1xi32>, %arg2: tensor<i32>, %arg3: tensor<i32>) -> tensor<2x4xi32> {
-// CHECK-NEXT:   %[[WORKLOAD0:.+]] = constant 8 : index
+// CHECK-NEXT:   [[WORKLOAD0:%.+]] = constant 8 : index
```
Comment (stellaraccident) [inline iree/compiler/Dialect/Flow/Transforms/test/transformation.mlir]: We're fighting with each other on % placement again

File: build_tools/cmake/iree_lit_test.cmake [hunk 9]
```diff
@@ -50,16 +50,18 @@ function(iree_lit_test)
     string(REPLACE "::" "_" _DATA_DEP_NAME ${_DATA_DEP})
     # TODO(*): iree_sh_binary so we can avoid this.
     if("${_DATA_DEP_NAME}" STREQUAL "iree_tools_IreeFileCheck")
-      list(APPEND _DATA_DEP_PATHS "${CMAKE_SOURCE_DIR}/iree/tools/IreeFileCheck.bat")
```
[the remark this answers is not in the document]
Comment (benvanik) [reply build_tools/cmake/iree_lit_test.cmake]: yeah, unfortunately all lit command lines run via bash so even on windows it uses the .sh :(

File: iree/compiler/Dialect/VM/IR/test/assignment_folding.mlir (hunk 6 above)
Comment (benvanik) [reply iree/compiler/Dialect/VM/IR/test/assignment_folding.mlir]: done

File: iree/compiler/Dialect/VM/Conversion/ImportUtils.cpp [hunk 10]
```diff
@@ -61,5 +63,148 @@ LogicalResult appendImportModule(StringRef importModuleSrc,
...
+  }
+  return prefix + llvm::utohexstr(static_cast<uint64_t>(
+                      llvm::hash_value(unsafeIdentifier)));
+}
+
+// Creates a module-level vm.rodata containing the string contents and returns
+// the dereferenced byte buffer.
+static Value createStringTableValue(Location loc, StringAttr attrValue,
+                                    Type inputType,
+                                    ConversionPatternRewriter &rewriter) {
```
[the remark this answers is not in the document]
Comment (benvanik) [reply iree/compiler/Dialect/VM/Conversion/ImportUtils.cpp]: agreed; here this is only called from conversion code below this that cannot run outside of conversion (and is under the Conversion/ path) so I feel like the rewriter makes it consistent - but for other util code outside of Conversion/ that's a good rule to follow

File: iree/hal/vulkan/direct_command_buffer.cc [hunk 11]
```diff
@@ -358,36 +376,82 @@ Status DirectCommandBuffer::CopyBuffer(Buffer* source_buffer,
...
+    DescriptorSet* descriptor_set,
+    absl::Span<const device_size_t> dynamic_offsets) {
+  IREE_TRACE_SCOPE0("DirectCommandBuffer::BindDescriptorSet");
+  ASSIGN_OR_RETURN(auto* device_executable_layout,
+                   CastExecutableLayout(executable_layout));
+  ASSIGN_OR_RETURN(auto* device_descriptor_set,
+                   CastDescriptorSet(descriptor_set));
+
+  // Vulkan takes uint32_t as the size here, unlike everywhere else.
+  absl::InlinedVector<uint32_t, 4> dynamic_offsets_i32(dynamic_offsets.size());
```
[the remark this answers is not in the document]
Comment (benvanik) [reply iree/hal/vulkan/direct_command_buffer.cc]: yeah :( it's a really hard maximum value to exceed but not impossible with sparse residency (https://www.asawicki.info/news_1698_vulkan_sparse_binding_-_a_quick_overview)
